sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,753 @@
+ initialised to zero. Note there is no guarantee that the memory allocation be
+ re-initialized to zero when it is re-allocated from the pool, so users must
+ re-initialize memory to zero if they wish for later allocations to have this
+ behavior.
+
+|`use_existing_memory`
+|The `use_existing_memory` property adds the requirement that the memory pool
+ will use an existing USM memory allocation provided by the user instead of
+ allocating from the SYCL runtime. This property takes a pointer to a valid USM
+ memory allocation of the same allocation kind as the memory pool is initialized
+ with and the size of that memory allocation. Using this property will
+ implicitly set the `maximum_size` and `initial_threshold` property to be that
+ of the size provided, and as such using the `maximum_size` or
+ `initial_threshold` properties in conjunction with this property will cause the
+ `memory_pool` constructor to throw an exception with the `errc::invalid` error
+ code.
+
+|===
+
+
+=== Default memory pools
+
+As well as being able to construct a memory pool explicitly, this extension
+introduces a default memory pool per device for each SYCL context and device
+pair for device allocations and a default memory pool per context for host
+allocations.
+
+New member functions are added to the `context` class to retrieve the default
+memory pool as a copy of the `memory_pool` object. This can be modified and have
+those modifications reflected as it conforms to the SYCL common reference
+semantics.
+
+[source,c++]
+----
+class context {
+
+  memory_pool context::ext_oneapi_get_default_memory_pool() const;
+
+  memory_pool context::ext_oneapi_get_default_memory_pool(device dev) const;
+
+}; // context
+----
+
+[source, c++]
+----
+memory_pool context::ext_oneapi_get_default_memory_pool() const;
+----
+
+_Returns_: The default memory pool associated with the context for allocating
+with the allocation kind `usm::alloc::host`.
+
+[source, c++]
+----
+memory_pool context::ext_oneapi_get_default_memory_pool(device dev) const;
+----
+
+_Returns_: The default memory pool associated with the context and `dev` for
+allocating with the allocation kind `usm::alloc::device`.
+
+
+=== Asynchronous malloc & free
+
+This extension introduces a series of new enqueue functions for enqueueing
+asynchronous malloc and free commands which operate with the memory pools also
+introduced in this extension.
+
+All enqueue functions introduced have overloads which take a SYCL `queue` and a
+SYCL `handler`. None of enqueue functions return a SYCL `event` directly, as
+this extension is in line with the
+link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
+  sycl_ext_oneapi_enqueue_functions] extension, so events are returned when
+calling `submit_with_event` and the `handler` overloads of these enqueue
+functions.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+
+void *async_malloc(queue q, size_t size);
```

Comment:
It seems that this API is only useful for in order queues. Should we change this entry point to also take a vector of `event`s?